### ISSUE TITLE
Issue About Constructing Preference Data (Webshop)

### ISSUE THREAD

**Issue by xuehui1991** [Yifan-Song793/ETO#7]
Hello Team,

I am currently working with the "construct preference" script and have a couple of queries that I believe would benefit from some clarification.

1. **`pred_traj_offset` Setting:**
   I noticed that for webshop data, [the `pred_traj_offset` is set to 10](https://github.com/Yifan-Song793/ETO/blob/3b67455376c60a5b65248a07707db54bc1c98b59/construct_preference.py#L40). Could you please explain the reasoning behind this specific value? 

2. **Reject Data Construction:**
   In the script, when constructing the reject data, I see the [following line of code](https://github.com/Yifan-Song793/ETO/blob/3b67455376c60a5b65248a07707db54bc1c98b59/construct_preference.py#L90):
   ```python
   rejected = pred_trajs[key][:len(golden_trajs[key]) + 5*2][1:]
   ```
   What is the purpose of this slicing operation, and why is the length of the golden trajectories plus `5*2` used? I am particularly interested in understanding the significance of the `5*2` multiplier and how it contributes to the rejection criteria.

Thank you in advance for your time and assistance. I am looking forward to your insights on these matters.

**Comment by Yifan-Song793** [comment]
Thanks for your response!

1. `pred_traj_offset` refers to the length of the task prompt and in-context example (including action & observation) in the generated trajectory. When constructing preference trajectories, the in-context example is not taken into account.
2. We add a slicing operation for the rejected trajectory. It is because some failed trajectories can be significantly longer than successful ones (e.g., repeatedly examing the temperature in ScienceWorld).  The specific `5*2` steps is just a heuristic strategy, and we haven't yet conducted an ablation study on it.

I'll be adding some clarifying comments in the `construct_preference.py` script to enhance understanding. Thanks again!

**Comment by xuehui1991** [comment]
Thank you for your quick response!

1. Regarding the first question on setting the `pred_traj_offset` for webshop data, if we don't use the in-context examples, could we just set the `pred_traj_offset` to 2 instead of 10?  As per the script, the pred_traj_offset is utilized at [line 69](https://github.com/Yifan-Song793/ETO/blob/3b67455376c60a5b65248a07707db54bc1c98b59/construct_preference.py#L69C32-L69C38), which extracts results from the exploration conversation except for the instruction_part.

2. For the Webshop dataset, it seems there isn't a significant occurrence of excessively long conversations. Therefore, I am wondering if we could handle it similarly to how another dataset is processed at [line 154](https://github.com/Yifan-Song793/ETO/blob/3b67455376c60a5b65248a07707db54bc1c98b59/construct_preference.py#L154). This approach could potentially streamline the process and align with the characteristics of the webshop data.

I appreciate your expertise and would love to hear your thoughts on this matter. Thank you.

**Comment by Yifan-Song793** [comment]
1. If you don't use the in-context examples, you can simply set the `pred_traj_offset` to 2. This value is determined by the length of task prompt and ICL examples before the actual task starts.
2. In the case of WebShop, you can safely omit the slicing operation. It works only for ScienceWorld in our experiments.

Feel free to reach out with any questions!

**Comment by xuehui1991** [comment]
Thanks @Yifan-Song793  , I truly appreciate your response. 

However, I have a few lingering questions regarding the current implementation on the main branch"

Is it code on the main branch does not include in-context examples when constructing the preference data? When I attempt to run run_eto.sh on the Webshop with the code directly from the main branch, it raises errors due to the two issues I mentioned earlier.

In an effort to address these issues, I made some modifications to the code as I proposed. Unfortunately, I have been unable to reproduce the results shown in Figure 4 of the paper.

I would like to seek your guidance on whether I might have missed something crucial, such as the inclusion of in-context examples, or perhaps there are parameters that differ from those set in run_eto.sh when you ran your experiments?

My current experimental results show the following:

After SFT, the model's performance on the test set is avg reward=0.6392.
After SFT + 1 iteration of ETO, the performance is avg reward=0.6399.
However, Figure 4 and Table 2 in the paper suggest that after SFT plus one iteration of ETO, the expected result should be avg reward=~0.67.

I am wondering if there is something significant that I have overlooked. Could you please provide any insights or suggestions that might help me align my results with those presented in the paper?

Thank you very much for your time and assistance.

**Comment by xuehui1991** [comment]
@Yifan-Song793

**Comment by sharptcode** [comment]
@xuehui1991 May I know what hyperparameters you used in SFT training? After the default SFT training with llama2-7b-chat, the performance on webshop has avg reward=0.57, way below 0.6392 you got or the paper reported. 

```
batch_size=64
num_train_epochs=3
learning_rate=2e-5
weight_decay=0.
```

**Comment by xuehui1991** [comment]
Apologies for the delayed reply. I did use the parameters in run_eto.sh, which means as the same parameters you mention in SFT training.  And I can reproduce the SFT result as paper.

Regarding the ETO outcome (SFT + rollout data + DPO training), I found the issue was identified in a recent version of transformers. train_dpo.py] (https://github.com/Yifan-Song793/ETO/blob/main/fastchat/train/train_dpo.py#L281), specifically line 281 and 289, if the ditype parameter isn't provided during model initialization, it could lead to unpredictable behavior during training. This doesn't trigger an error but generates a warning. I've addressed this and am now able to reproduce the results. Thank you for your assistance.
